vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,744 @@
+
+    pI2cDescriptor->xCallback = NULL;
+    pI2cDescriptor->pvUserContext = NULL;
+    pI2cDescriptor->usSlaveAddr = _UNSET_SLAVE_ADDRESS;
+    pI2cDescriptor->usTransmittedTxBytes = 0;
+    pI2cDescriptor->usReceivedRxBytes = 0;
+    pI2cDescriptor->flags = _FLAG_INITIALIZER;
+    pI2cDescriptor->xConfig.ulBusFreq = 0;
+    pI2cDescriptor->xConfig.ulMasterTimeout = 0;
+
+    vSemaphoreDelete( pI2cDescriptor->xSemphr );
+
+    /* HAL_I2C_DeInit always returns HAL_OK as long as the handle is not NULL. */
+    return IOT_I2C_SUCCESS;
+}
+
+/*-----------------------------------------------------------*/
+
+/*
+ * Actual ISR registered in vector table.
+ */
+void I2C1_EV_IRQHandler( void )
+{
+    HAL_I2C_EV_IRQHandler( &( _i2cContexts[ 0 ].xHandle ) );
+}
+
+void I2C1_ER_IRQHandler( void )
+{
+    HAL_I2C_ER_IRQHandler( &( _i2cContexts[ 0 ].xHandle ) );
+}
+
+void I2C2_EV_IRQHandler( void )
+{
+    HAL_I2C_EV_IRQHandler( &( _i2cContexts[ 1 ].xHandle ) );
+}
+
+void I2C2_ER_IRQHandler( void )
+{
+    HAL_I2C_ER_IRQHandler( &( _i2cContexts[ 1 ].xHandle ) );
+}
+
+void I2C3_EV_IRQHandler( void )
+{
+    HAL_I2C_EV_IRQHandler( &( _i2cContexts[ 2 ].xHandle ) );
+}
+
+void I2C3_ER_IRQHandler( void )
+{
+    HAL_I2C_ER_IRQHandler( &( _i2cContexts[ 2 ].xHandle ) );
+}
+
+/*-------------ST API Implementation Overrides---------------*/
+
+/**
+ * @brief Initializes low level hardware
+ */
+void HAL_I2C_MspInit( I2C_HandleTypeDef * hi2c )
+{
+    GPIO_InitTypeDef xGpio_init_structure;
+
+    /* Configure the GPIOs */
+    DISCOVERY_I2Cx_SCL_SDA_GPIO_CLK_ENABLE();
+
+    /* Configure I2C Tx, Rx as alternate function */
+    xGpio_init_structure.Pin = DISCOVERY_I2Cx_SCL_PIN | DISCOVERY_I2Cx_SDA_PIN;
+    xGpio_init_structure.Mode = GPIO_MODE_AF_OD;
+    xGpio_init_structure.Pull = GPIO_PULLUP;
+    xGpio_init_structure.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+    xGpio_init_structure.Alternate = DISCOVERY_I2Cx_SCL_SDA_AF;
+    HAL_GPIO_Init( DISCOVERY_I2Cx_SCL_SDA_GPIO_PORT, &xGpio_init_structure );
+
+    /* Configure the I2C peripheral */
+    DISCOVERY_I2Cx_CLK_ENABLE();
+
+    DISCOVERY_I2Cx_FORCE_RESET();
+
+    DISCOVERY_I2Cx_RELEASE_RESET();
+
+    /* Enable and set I2Cx Interrupt to a lower priority */
+    HAL_NVIC_SetPriority( DISCOVERY_I2Cx_EV_IRQn, 0x0F, 0 );
+    HAL_NVIC_EnableIRQ( DISCOVERY_I2Cx_EV_IRQn );
+}
+
+void HAL_I2C_MspDeInit( I2C_HandleTypeDef * hi2c )
+{
+    GPIO_InitTypeDef xGpio_init_structure;
+
+    /* Configure I2C Tx, Rx as alternate function */
+    xGpio_init_structure.Pin = DISCOVERY_I2Cx_SCL_PIN | DISCOVERY_I2Cx_SDA_PIN;
+    HAL_GPIO_DeInit( DISCOVERY_I2Cx_SCL_SDA_GPIO_PORT, xGpio_init_structure.Pin );
+
+    DISCOVERY_I2Cx_SCL_SDA_GPIO_CLK_DISABLE();
+
+    DISCOVERY_I2Cx_CLK_DISABLE();
+}
+
+void HAL_I2C_MasterTxCpltCallback( I2C_HandleTypeDef * hi2c )
+{
+    IotI2CDescriptor_t * pDescriptor = _toDescriptor( hi2c );
+    BaseType_t higherPriorityTaskWoken = pdFALSE;
+
+    /* Subtract not transferred yet. */
```

Comment:
This is confusing. We're in a transmit-complete callback, but we're decreasing the number of bytes that have been transferred. Why?

Reply:
Because XferCount is NOT the number of bytes that have been transferred. It is the number of bytes still left, not yet transferred. That is just how ST driver tracks state.